File: backend/services/reglas_negocio.py
"""
Motor de Reglas de Negocio — Selección automática de plantilla y validaciones.

Carga reglas de manera declarativa desde data/rules_config.json y data/templates_config.json.
"""
import os
import json
from typing import List, Dict, Optional
from backend.config import BASE_DIR

RULES_PATH = os.path.join(BASE_DIR, "data", "rules_config.json")
TEMPLATES_PATH = os.path.join(BASE_DIR, "data", "templates_config.json")

PLANTILLAS = [
    "CO BASE", "CO BASE (2)", "CO BASE (3)", "CO (3)", "CO X",
    "CO+mov+teb", "CO TP", "CO GEF", "CO sin eps", "CO Chofer",
    "CO maternidad", "PRACTICANTE PRE", "PRACTICANTE PRO",
    "CSIR", "PART TIME", "PART TIME (2)", "operador", "PREPARADOR FISICO",
]


def cargar_reglas() -> list:
    if os.path.exists(RULES_PATH):
        try:
            with open(RULES_PATH, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception:
            pass
    return []


def cargar_plantillas() -> dict:
    if os.path.exists(TEMPLATES_PATH):
        try:
            with open(TEMPLATES_PATH, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception:
            pass
    return {}


def _evaluar_condicion(cond: dict, solicitud: dict) -> bool:
    if "any" in cond:
        return any(_evaluar_condicion(c, solicitud) for c in cond["any"])
    if "all" in cond:
        return all(_evaluar_condicion(c, solicitud) for c in cond["all"])
    
    field = cond.get("field")
    operator = cond.get("operator", "equal")
    target_val = cond.get("value")
    
    raw_val = solicitud.get(field)
    
    # Normalización para comparaciones de texto
    val_str = str(raw_val or "").upper().strip()
    target_str = str(target_val or "").upper().strip()
    
    if operator == "contains":
        return target_str in val_str
    elif operator == "equal":
        return val_str == target_str
    elif operator == "not_equal":
        return val_str != target_str
    elif operator == "true":
        return bool(raw_val) is True
    elif operator == "false":
        return bool(raw_val) is False
    
    return False


def determinar_plantilla(solicitud: dict) -> str:
    """
    Determina automáticamente la plantilla de carta oferta evaluando las reglas
    cargadas desde rules_config.json.
    """
    rules = cargar_reglas()
    if not rules:
        return determinar_plantilla_fallback(solicitud)
        
    for rule in rules:
        plantilla = rule.get("plantilla")
        condiciones = rule.get("condiciones", [])
        
        if not condiciones or all(_evaluar_condicion(c, solicitud) for c in condiciones):
            return plantilla
            
    return "CO BASE"


def determinar_plantilla_fallback(solicitud: dict) -> str:
    """Fallback por código si no existe el archivo JSON."""
    modalidad = (solicitud.get("modalidad") or "").upper().strip()
    tipo_ingreso = (solicitud.get("tipo_ingreso") or "").upper().strip()
    unidad = (solicitud.get("unidad") or "").upper().strip()
    incluye_eps = solicitud.get("incluye_eps", True)
    incluye_movilidad = solicitud.get("incluye_movilidad", False)
    incluye_tarjeta = solicitud.get("incluye_tarjeta_alimentos", False)
    incluye_transporte = solicitud.get("incluye_bono_transporte", False)
    tipo_personal = (solicitud.get("tipo_personal") or "").upper().strip()
    jornada = solicitud.get("jornada", "48")
    puesto = (solicitud.get("puesto") or "").upper().strip()
    categoria_trab = (solicitud.get("categoria_trabajador") or "").upper().strip()

    if "PRACTICANTE" in modalidad or "PRACTICANTE" in puesto or "PRACTICANTE" in tipo_ingreso:
        if "PRE" in modalidad or "PRE" in puesto:
            return "PRACTICANTE PRE"
        return "PRACTICANTE PRO"

    if "MATERNIDAD" in tipo_ingreso or "LICENCIA" in tipo_ingreso:
        return "CO maternidad"

    if "PREPARADOR" in puesto or "INSTRUCTOR" in puesto:
        return "PREPARADOR FISICO"

    if "OPERADOR" in puesto and "PROMOTOR" in puesto:
        return "operador"

    if unidad == "GEF" or unidad == "IE":
        return "CO GEF"

    if unidad == "CSIR":
        return "CSIR"

    if "PART" in modalidad or jornada != "48":
        if incluye_eps:
            return "PART TIME"
        return "PART TIME (2)"

    if "CHOFER" in puesto:
        return "CO Chofer"

    if incluye_movilidad and incluye_tarjeta:
        return "CO+mov+teb"
    if incluye_transporte and incluye_tarjeta:
        return "CO+mov+teb"

    if not incluye_eps:
        return "CO sin eps"

    if "DOCENTE" in puesto and "TIEMPO PARCIAL" in modalidad:
        return "CO TP"

    if "GERENTE" in puesto or "INDETERMINADO" in tipo_personal:
        return "CO X"

    if "DIRECCION" in tipo_personal and "NO FISCALIZABLE" in categoria_trab:
        return "CO (3)"

    if "CONFIANZA" in tipo_personal and "NO FISCALIZABLE" in categoria_trab:
        return "CO BASE (2)"

    if "SUJETO" in categoria_trab:
        return "CO BASE (3)"

    return "CO BASE"


def validar_solicitud(solicitud: dict) -> List[str]:
    """
    Validaciones que previenen errores ANTES de generar la carta.
    Devuelve lista de errores (vacía si todo está OK).
    """
    errores = []

    # Campos obligatorios
    if not solicitud.get("nombres_apellidos"):
        errores.append("El nombre del colaborador es obligatorio")
    if not solicitud.get("puesto"):
        errores.append("Debe especificar el puesto")
    if not solicitud.get("unidad"):
        errores.append("Debe seleccionar la unidad de negocio")

    # Salario
    salario = solicitud.get("salario")
    if salario is not None and salario <= 0:
        errores.append("El salario debe ser mayor a 0")

    # DNI
    dni = solicitud.get("dni", "")
    if dni and len(str(dni).strip()) not in (8, 9, 12):
        errores.append("El DNI debe tener 8 dígitos o CEX 9-12 caracteres")

    # Reemplazo requiere datos del saliente
    tipo_ingreso = (solicitud.get("tipo_ingreso") or "").upper()
    if "REEMPLAZO" in tipo_ingreso:
        if not solicitud.get("codigo_reemplazo"):
            errores.append("Para reemplazos, debe indicar el código del saliente")
        if not solicitud.get("nombre_reemplazo"):
            errores.append("Para reemplazos, debe indicar el nombre del saliente")

    # CECO
    if not solicitud.get("codigo_cr"):
        errores.append("El código CR (CECO) es obligatorio")

    # Fechas
    if not solicitud.get("fecha_tentativa_ingreso"):
        errores.append("La fecha tentativa de ingreso es obligatoria")

    return errores


def obtener_config_plantilla(nombre: str) -> dict:
    """Devuelve la configuración automática de una plantilla."""
    templates = cargar_plantillas()
    if nombre in templates:
        config = templates[nombre].get("config_defecto", {})
        jornada = config.get("jornada", "48")
        incluye_eps = config.get("incluye_eps", True)
        periodo_prueba = config.get("periodo_prueba", "03 meses")
        texto_jornada = (
            "tiempo parcial" if jornada == "parcial"
            else "treinta (30) horas semanales" if jornada == "30"
            else "cuarenta y ocho (48) horas semanales"
        )
        return {
            "jornada": jornada,
            "incluye_eps": incluye_eps,
            "periodo_prueba": periodo_prueba,
            "texto_jornada": texto_jornada,
            "incluye_tarjeta_alimentos": config.get("incluye_tarjeta_alimentos", False),
            "monto_tarjeta_alimentos": config.get("monto_tarjeta_alimentos"),
            "incluye_bono_transporte": config.get("incluye_bono_transporte", False),
            "monto_bono_transporte": config.get("monto_bono_transporte"),
        }

    # Fallback clásico
    return {
        "jornada": "48", "incluye_eps": True, "periodo_prueba": "03 meses",
        "texto_jornada": "cuarenta y ocho (48) horas semanales",
    }

File: frontend-react/src/services/reglasNegocio.ts
// Declarative business rules engine for letter template determination and validation
import { buildAssetUrl } from './paths';

export interface Condicion {
  field: string;
  operator?: 'equal' | 'not_equal' | 'contains' | 'true' | 'false';
  value?: any;
  any?: Condicion[];
  all?: Condicion[];
}

export interface ReglaNegocio {
  plantilla: string;
  condiciones: Condicion[];
}

export interface PlantillaConfig {
  config_defecto: {
    jornada?: string;
    incluye_eps?: boolean;
    periodo_prueba?: string;
    incluye_tarjeta_alimentos?: boolean;
    monto_tarjeta_alimentos?: number;
    incluye_bono_transporte?: boolean;
    monto_bono_transporte?: number;
  };
  variables: Record<string, string>;
  parrafos: string[];
}

let reglas: ReglaNegocio[] = [];
let plantillas: Record<string, PlantillaConfig> = {};
let isConfigLoaded = false;

export const loadBusinessRules = async (): Promise<void> => {
  if (isConfigLoaded) return;
  try {
    const [reglasRes, plantillasRes] = await Promise.all([
      fetch(buildAssetUrl('data/config/rules_config.json')),
      fetch(buildAssetUrl('data/config/templates_config.json'))
    ]);

    if (reglasRes.ok) reglas = await reglasRes.json();
    if (plantillasRes.ok) plantillas = await plantillasRes.json();
    isConfigLoaded = true;
  } catch (error) {
    console.warn("Could not load rules from API, using fallback rules.", error);
    // Load some basic fallback configuration
    reglas = [];
    plantillas = {};
    isConfigLoaded = false;
  }
};

const evaluarCondicion = (cond: Condicion, solicitud: any): boolean => {
  if (cond.any) {
    return cond.any.some((c) => evaluarCondicion(c, solicitud));
  }
  if (cond.all) {
    return cond.all.every((c) => evaluarCondicion(c, solicitud));
  }

  const { field, operator = 'equal', value: targetVal } = cond;
  const rawVal = solicitud[field];
  
  const valStr = String(rawVal ?? '').toUpperCase().trim();
  const targetStr = String(targetVal ?? '').toUpperCase().trim();

  switch (operator) {
    case 'contains':
      return valStr.includes(targetStr);
    case 'equal':
      return valStr === targetStr;
    case 'not_equal':
      return valStr !== targetStr;
    case 'true':
      return !!rawVal === true;
    case 'false':
      return !!rawVal === false;
    default:
      return false;
  }
};

export const determinarPlantillaCliente = async (solicitud: any): Promise<string> => {
  await loadBusinessRules();
  
  if (reglas.length === 0) {
    return determinarPlantillaFallback(solicitud);
  }

  for (const rule of reglas) {
    const condiciones = rule.condiciones || [];
    if (condiciones.length === 0 || condiciones.every(c => evaluarCondicion(c, solicitud))) {
      return rule.plantilla;
    }
  }

  return "CO BASE";
};

// Fallback logic matching backend/services/reglas_negocio.py
const determinarPlantillaFallback = (solicitud: any): string => {
  const modalidad = String(solicitud.modalidad || '').toUpperCase().trim();
  const tipoIngreso = String(solicitud.tipo_ingreso || '').toUpperCase().trim();
  const unidad = String(solicitud.unidad || '').toUpperCase().trim();
  const incluyeEps = solicitud.incluye_eps !== false;
  const incluyeMovilidad = !!solicitud.incluye_movilidad;
  const incluyeTarjeta = !!solicitud.incluye_tarjeta_alimentos;
  const incluyeTransporte = !!solicitud.incluye_bono_transporte;
  const tipoPersonal = String(solicitud.tipo_personal || '').toUpperCase().trim();
  const jornada = String(solicitud.jornada || '48');
  const puesto = String(solicitud.puesto || '').toUpperCase().trim();
  const categoriaTrab = String(solicitud.categoria_trabajador || '').toUpperCase().trim();

  if (modalidad.includes('PRACTICANTE') || puesto.includes('PRACTICANTE') || tipoIngreso.includes('PRACTICANTE')) {
    if (modalidad.includes('PRE') || puesto.includes('PRE')) {
      return "PRACTICANTE PRE";
    }
    return "PRACTICANTE PRO";
  }

  if (tipoIngreso.includes('MATERNIDAD') || tipoIngreso.includes('LICENCIA')) {
    return "CO maternidad";
  }

  if (puesto.includes('PREPARADOR') || puesto.includes('INSTRUCTOR')) {
    return "PREPARADOR FISICO";
  }

  if (puesto.includes('OPERADOR') && puesto.includes('PROMOTOR')) {
    return "operador";
  }

  if (unidad === 'GEF' || unidad === 'IE') return "CO GEF";
  if (unidad === 'CSIR') return "CSIR";

  if (modalidad.includes('PART') || jornada !== '48') {
    return incluyeEps ? "PART TIME" : "PART TIME (2)";
  }

  if (puesto.includes('CHOFER')) return "CO Chofer";

  if ((incluyeMovilidad && incluyeTarjeta) || (incluyeTransporte && incluyeTarjeta)) {
    return "CO+mov+teb";
  }

  if (!incluyeEps) return "CO sin eps";

  if (puesto.includes('DOCENTE') && modalidad.includes('TIEMPO PARCIAL')) {
    return "CO TP";
  }

  if (puesto.includes('GERENTE') || tipoPersonal.includes('INDETERMINADO')) {
    return "CO X";
  }

  if (tipoPersonal.includes('DIRECCION') && categoriaTrab.includes('NO FISCALIZABLE')) {
    return "CO (3)";
  }

  if (tipoPersonal.includes('CONFIANZA') && categoriaTrab.includes('NO FISCALIZABLE')) {
    return "CO BASE (2)";
  }

  if (categoriaTrab.includes('SUJETO')) {
    return "CO BASE (3)";
  }

  return "CO BASE";
};

export const getTemplateDetails = async (nombre: string): Promise<PlantillaConfig | null> => {
  await loadBusinessRules();
  return plantillas[nombre] || null;
};
